code/modules/paperwork/paper.dm

```diff
@@ -302,103 +244,102 @@
+		var/obj/item/pen/PEN = O
+		data["pen_font"] = PEN.font
+		data["pen_color"] = PEN.colour
+		data["edit_mode"] = MODE_WRITING
+		data["is_crayon"] = FALSE
+		data["stamp_class"] = "FAKE"
+		data["stamp_icon_state"] = "FAKE"
+	else if(istype(O, /obj/item/stamp))
+		var/datum/asset/spritesheet/sheet = get_asset_datum(/datum/asset/spritesheet/simple/paper)
+		data["stamp_icon_state"] = O.icon_state
+		data["stamp_class"] = sheet.icon_class_name(O.icon_state)
+		data["edit_mode"] = MODE_STAMPING
+		data["pen_font"] = "FAKE"
+		data["pen_color"] = "FAKE"
+		data["is_crayon"] = FALSE
+	else
+		data["edit_mode"] = MODE_READING
+		data["pen_font"] = "FAKE"
+		data["pen_color"] = "FAKE"
+		data["is_crayon"] = FALSE
+		data["stamp_icon_state"] = "FAKE"
+		data["stamp_class"] = "FAKE"
 	data["field_counter"] = field_counter
 	data["form_fields"] = form_fields
 
 	return data
 
-/obj/item/paper/ui_act(action, params, datum/tgui/ui, datum/ui_state/default/paper_state/state)
+/obj/item/paper/ui_act(action, params,datum/tgui/ui)
 	if(..())
 		return
 	switch(action)
 		if("stamp")
 			var/stamp_x = text2num(params["x"])
 			var/stamp_y = text2num(params["y"])
 			var/stamp_r = text2num(params["r"])	// rotation in degrees
-
+			var/stamp_icon_state = params["stamp_icon_state"]
+			var/stamp_class = params["stamp_class"]
 			if (isnull(stamps))
-				stamps = new/list()
+				stamps = list()
 			if(stamps.len < MAX_PAPER_STAMPS)
 				// I hate byond when dealing with freaking lists
-				stamps += list(list(state.stamp_class, stamp_x,  stamp_y,stamp_r))	/// WHHHHY
+				stamps[++stamps.len] = list(stamp_class, stamp_x, stamp_y, stamp_r)	/// WHHHHY
 
 				/// This does the overlay stuff
 				if (isnull(stamped))
-					stamped = new/list()
+					stamped = list()
 				if(stamped.len < MAX_PAPER_STAMPS_OVERLAYS)
-					var/mutable_appearance/stampoverlay = mutable_appearance('icons/obj/bureaucracy.dmi', "paper_[state.stamp_icon_state]")
+					var/mutable_appearance/stampoverlay = mutable_appearance('icons/obj/bureaucracy.dmi', "paper_[stamp_icon_state]")
 					stampoverlay.pixel_x = rand(-2, 2)
 					stampoverlay.pixel_y = rand(-3, 2)
 					add_overlay(stampoverlay)
-					LAZYADD(stamped, state.stamp_icon_state)
+					LAZYADD(stamped, stamp_icon_state)
 
-				ui.user.visible_message("<span class='notice'>[ui.user] stamps [src] with [state.stamp_name]!</span>", "<span class='notice'>You stamp [src] with [state.stamp_name]!</span>")
+				update_static_data(usr,ui)
+				ui.user.visible_message("<span class='notice'>[ui.user] stamps [src] with [stamp_class]!</span>", "<span class='notice'>You stamp [src] with [stamp_class]!</span>")
 			else
 				to_chat(usr, pick("You try to stamp but you miss!", "There is no where else you can stamp!"))
-
-			update_all_ui()
 			. = TRUE
 
 		if("save")
 			var/in_paper = params["text"]
 			var/paper_len = length(in_paper)
-			var/list/fields = params["form_fields"]
+			// var/list/fields = params["form_fields"]
```

Comment:
Leftover? Remove commented out code if no longer relevant.

Reply:
Yea removed it.  At some point, maybe we want some device to read forms that have been filled out but much more work would have to be done.